<launch>

    <arg name="publish_vehicle_state_from_xsens" default="true" />

    <!-- parameters -->
    <!--arg name="device" default="auto" doc="device file of the IMU"/ --> <!-- Use this if you are running on DrivePX -->
    <arg name="device" default="/dev/ttyUSB.xsens" doc="device file of the IMU"/> <!-- Use this if you are running on MSI -->
    <arg name="baudrate" default="0" doc="baudrate of the IMU"/>
    <arg name="timeout" default="0.002" doc="timeout for the IMU communication"/>
    <arg name="frame_id" default="/imu" doc="frame id of the IMU"/>
    <arg name="frame_local" default="ENU" doc="desired frame orientation (ENU, NED or NWU)"/>
    <arg name="no_rotation_duration" default="0" doc="duration (int in seconds) of the no-rotation calibration procedure"/>
    <arg name="angular_velocity_covariance_diagonal" default="[0.0004, 0.0004, 0.0004]" doc="Diagonal elements of angular velocity covariance matrix"/>
    <arg name="linear_acceleration_covariance_diagonal" default="[0.0004, 0.0004, 0.0004]" doc="Diagonal elements of linear acceleration covariance matrix"/>
    <arg name="orientation_covariance_diagonal" default="[0.01745, 0.01745, 0.15708]" doc="Diagonal elements of the orientation covariance matrix"/>

    <!-- node -->
    <group ns="xsens">
      <node pkg="xsens_driver" type="mtnode.py" name="xsens_driver" output="screen" respawn="true" respawn_delay="5">
        <param name="device" value="$(arg device)"/>
        <param name="baudrate" value="$(arg baudrate)"/>
        <param name="timeout" value="$(arg timeout)"/>
        <param name="frame_id" value="$(arg frame_id)"/>
        <param name="frame_local" value="$(arg frame_local)"/>
        <param name="no_rotation_duration" value="$(arg no_rotation_duration)"/>
        <rosparam param="angular_velocity_covariance_diagonal" subst_value="True">$(arg angular_velocity_covariance_diagonal)</rosparam>
        <rosparam param="linear_acceleration_covariance_diagonal" subst_value="True">$(arg linear_acceleration_covariance_diagonal)</rosparam>
        <rosparam param="orientation_covariance_diagonal" subst_value="True">$(arg orientation_covariance_diagonal)</rosparam>
      </node>
    </group>

    <group if="$(arg publish_vehicle_state_from_xsens)" >
        <node pkg="xsens_driver" type="hri_xsens_wrapper_node" name="hri_xsens_wrapper_node" output="screen">
          <param name="gps_output_topic" value="/xsens_gps_reading"/>
          <param name="xsens_ns" value="/xsens"/>
        </node>

        <node pkg="gps2hriframe" type="gps2hriframe_node" name="gps2hriframe_node_xsens" output="screen" >
            <param name="lat_ref" value="37.390" type="double" />
            <param name="lon_ref" value="-122.052" type="double" />
            <param name="hea_ref" value="310.75" type="double" />
            <param name="alt_ref" value="0.0" type="double" />
            <param name="veh_state_rate" value="25.0" type="double" />
            <param name="base_link_tf_rate" value="25.0" type="double" />
            <param name="offset_x" value="0.4" type="double" />
            <param name="offset_y" value="0.0" type="double" />
            <param name="offset_z" value="-0.4" type="double" />
            <param name="head_north_east" value="true" type="bool" />
            <param name="topic_name_gps" value="xsens_gps_reading" />
            <param name="topic_name_veh_state" value="vehicle_state_xsens" />
            <param name="tf_name_gps" value="base_link_xsens"/>
        </node>
    </group>
</launch>
